src/Distribution/Extra/Doctest.hs

```diff
@@ -165,88 +172,107 @@ generateBuildModule testSuiteName flags pkg lbi = do
+           let compAutogenDir = autogenModulesDir lbi
 #endif
 
-    -- Lib sources and includes
-    iArgs' <- mapM (fmap ("-i"++) . makeAbsolute)
-        $ libAutogenDir            -- autogenerated files
-        : (distPref ++ "/build")   -- preprocessed files (.hsc -> .hs); "build" is hardcoded in Cabal.
-        : hsSourceDirs libBI
-    includeArgs <- mapM (fmap ("-I"++) . makeAbsolute) $ includeDirs libBI
-    -- We clear all includes, so the CWD isn't used.
-    let iArgs = "-i" : iArgs'
+           -- Lib sources and includes
+           iArgs' <- mapM (fmap ("-i"++) . makeAbsolute)
+               $ compAutogenDir           -- autogenerated files
+               : (distPref ++ "/build")   -- preprocessed files (.hsc -> .hs); "build" is hardcoded in Cabal.
+               : hsSourceDirs compBI
+           includeArgs <- mapM (fmap ("-I"++) . makeAbsolute) $ includeDirs compBI
+           -- We clear all includes, so the CWD isn't used.
+           let iArgs = "-i" : iArgs'
 
-    -- default-extensions
-    let extensionArgs = map (("-X"++) . display) $ defaultExtensions libBI
+           -- default-extensions
+           let extensionArgs = map (("-X"++) . display) $ defaultExtensions compBI
 
-    -- CPP includes, i.e. include cabal_macros.h
-    let cppFlags = map ("-optP"++) $
-            [ "-include", libAutogenDir ++ "/cabal_macros.h" ]
-            ++ cppOptions libBI
+           -- CPP includes, i.e. include cabal_macros.h
+           let cppFlags = map ("-optP"++) $
+                   [ "-include", compAutogenDir ++ "/cabal_macros.h" ]
+                   ++ cppOptions compBI
 
-    withTestLBI pkg lbi $ \suite suitecfg -> when (testName suite == fromString testSuiteName) $ do
-      let testBI = testBuildInfo suite
+           let testBI = testBuildInfo suite
 
-      -- TODO: `words` is not proper parser (no support for quotes)
-      let additionalFlags = maybe [] words
-            $ lookup "x-doctest-options"
-            $ customFieldsBI testBI
+           -- TODO: `words` is not proper parser (no support for quotes)
+           let additionalFlags = maybe [] words
+                 $ lookup "x-doctest-options"
+                 $ customFieldsBI testBI
 
-      let additionalModules = maybe [] words
-            $ lookup "x-doctest-modules"
-            $ customFieldsBI testBI
+           let additionalModules = maybe [] words
+                 $ lookup "x-doctest-modules"
+                 $ customFieldsBI testBI
 
-      let additionalDirs' = maybe [] words
-            $ lookup "x-doctest-source-dirs"
-            $ customFieldsBI testBI
-      additionalDirs <- mapM (fmap ("-i" ++) . makeAbsolute) additionalDirs'
+           let additionalDirs' = maybe [] words
+                 $ lookup "x-doctest-source-dirs"
+                 $ customFieldsBI testBI
+           additionalDirs <- mapM (fmap ("-i" ++) . makeAbsolute) additionalDirs'
 
+           let compSuffix          = maybe "" ("_" ++) (mbCompName comp)
```

Comment:
component name may be invalid identifier, e.g in *can* contain dash, which isn't too far fetched.

Reply:
Good catch. I'm now applying the same `fixchar` trick that `Cabal` does when generating `MIN_VERSION_foo_bar` macros (for a package `foo-bar`).